--- ../linux-4.9.100/include/net/flow_dissector.h	2018-05-16 03:08:45.000000000 -0500
+++ ./include/net/flow_dissector.h	2018-09-18 22:07:50.902055195 -0500
@@ -28,7 +28,7 @@
 struct flow_dissector_key_basic {
 	__be16	n_proto;
 	u8	ip_proto;
-	u8	padding;
+	u8	padding;   //not used space, to store the dev id by Kun
 };

 struct flow_dissector_key_tags {
@@ -128,10 +128,11 @@
 	FLOW_DISSECTOR_KEY_FLOW_LABEL, /* struct flow_dissector_key_flow_tags */
 	FLOW_DISSECTOR_KEY_GRE_KEYID, /* struct flow_dissector_key_keyid */
 	FLOW_DISSECTOR_KEY_MPLS_ENTROPY, /* struct flow_dissector_key_keyid */
-
+
 	FLOW_DISSECTOR_KEY_MAX,
 };

+
 #define FLOW_DISSECTOR_F_PARSE_1ST_FRAG		BIT(0)
 #define FLOW_DISSECTOR_F_STOP_AT_L3		BIT(1)
 #define FLOW_DISSECTOR_F_STOP_AT_FLOW_LABEL	BIT(2)
