--- ../linux-4.9.100/include/net/flow_dissector.h	2018-05-16 03:08:45.000000000 -0500
+++ ./include/net/flow_dissector.h	2018-09-19 12:03:34.401621779 -0500
@@ -28,7 +28,7 @@
 struct flow_dissector_key_basic {
 	__be16	n_proto;
 	u8	ip_proto;
-	u8	padding;
+	u8	padding;   //not used space, to store the dev id by Kun
 };

 struct flow_dissector_key_tags {
@@ -128,10 +128,11 @@
 	FLOW_DISSECTOR_KEY_FLOW_LABEL, /* struct flow_dissector_key_flow_tags */
 	FLOW_DISSECTOR_KEY_GRE_KEYID, /* struct flow_dissector_key_keyid */
 	FLOW_DISSECTOR_KEY_MPLS_ENTROPY, /* struct flow_dissector_key_keyid */
-
+
 	FLOW_DISSECTOR_KEY_MAX,
 };

+
 #define FLOW_DISSECTOR_F_PARSE_1ST_FRAG		BIT(0)
 #define FLOW_DISSECTOR_F_STOP_AT_L3		BIT(1)
 #define FLOW_DISSECTOR_F_STOP_AT_FLOW_LABEL	BIT(2)
@@ -149,14 +150,16 @@
 };

 struct flow_keys {
-	struct flow_dissector_key_control control;
+	struct flow_dissector_key_control control;  //64 bits
 #define FLOW_KEYS_HASH_START_FIELD basic
-	struct flow_dissector_key_basic basic;
-	struct flow_dissector_key_tags tags;
-	struct flow_dissector_key_vlan vlan;
-	struct flow_dissector_key_keyid keyid;
-	struct flow_dissector_key_ports ports;
-	struct flow_dissector_key_addrs addrs;
+	struct flow_dissector_key_basic basic;      //32
+
+	struct flow_dissector_key_tags tags;        //32
+	struct flow_dissector_key_vlan vlan;        //32
+	struct flow_dissector_key_keyid keyid;      //32
+
+	struct flow_dissector_key_ports ports;      //32
+	struct flow_dissector_key_addrs addrs;      //64
 };

 #define FLOW_KEYS_HASH_OFFSET		\
