--- ../linux-4.9.100/include/linux/skbuff.h	2018-05-16 03:08:45.000000000 -0500
+++ ./include/linux/skbuff.h	2018-09-20 20:56:35.405675649 -0500
@@ -23,6 +23,7 @@
 #include <linux/rbtree.h>
 #include <linux/socket.h>

+
 #include <linux/atomic.h>
 #include <asm/types.h>
 #include <linux/spinlock.h>
@@ -1139,7 +1140,8 @@

 static inline __u32 skb_get_hash(struct sk_buff *skb)
 {
-	if (!skb->l4_hash && !skb->sw_hash)
+//add by Kun
+//	if (!skb->l4_hash && !skb->sw_hash)
 		__skb_get_hash(skb);
    
+ //@l4_hash: indicate hash is a canonical 4-tuple hash over transport ports.
+ //@sw_hash: indicates hash was computed in software stack
+ //sw_hash will be set to 1 at __skb_set_sw_hash when the skb is first time processed for hash

 	return skb->hash;
